client/src/components/ElementEditor/ElementList.js

```diff
@@ -11,6 +11,101 @@ import { getDragIndicatorIndex } from 'lib/dragHelpers';
 
 class ElementList extends Component {
+  constructor(props) {
+    super(props);
+    this.resetState = this.resetState.bind(this);
+    this.handleBeforeSubmitForm = this.handleBeforeSubmitForm.bind(this);
+    this.handleAfterSubmitResponse = this.handleAfterSubmitResponse.bind(this);
+    this.state = {
+      // saveAllElements will be set to true in entwine.js in the 'onbeforesubmitform' "hook"
+      // which is triggered by LeftAndMain submitForm()
+      saveAllElements: false,
+      // increment is also set in entwine.js in the 'onbeforesubmitform' "hook"
+      increment: 0,
+      hasUnsavedChangesBlockIDs: {},
+      validBlockIDs: {},
+    };
+    // Update the sharedObject so that setState() can be called from entwine.js
+    this.props.sharedObject.setState = this.setState.bind(this);
+  }
+
+  componentDidUpdate(prevProps, prevState) {
+    // Number of blocks just changed after graphql response
+    if (this.props.blocks !== prevProps.blocks) {
+      this.resetState(prevState, false);
+      return;
+    }
+    // Saving all elements and state has just updated because of a formSchema response from inline save
+    if (this.state.saveAllElements) {
+      const unsavedChangesBlockIDs = this.props.blocks
+        .map(block => parseInt(block.id, 10))
+        .filter(blockID => this.state.hasUnsavedChangesBlockIDs[blockID]);
+      const allValidated = unsavedChangesBlockIDs.every(blockID => this.state.validBlockIDs[blockID] !== null);
+      if (allValidated) {
+        const allValid = unsavedChangesBlockIDs.every(blockID => this.state.validBlockIDs[blockID]);
+        // entwineResolve is bound in entwine.js
+        this.props.sharedObject.entwineResolve(allValid);
+        this.resetState(prevState, allValid);
+        this.setState({ saveAllElements: false });
+      }
+    }
+  }
+
+  resetState(prevState, resetHasUnsavedChangesBlockIDs) {
+    if (this.props.blocks === null) {
+      return;
+    }
+    // validBlockIDs (though not hasUnsavedChangesBlockIDs) uses a tri-state
+    // - null: not saved
+    // - true: saved, valid
+    // - false: attempted save, invalid
+    const hasUnsavedChangesBlockIDs = {};
+    const validBlockIDs = {};
+    this.props.blocks.forEach(block => {
+      const blockID = parseInt(block.id, 10);
+      if (resetHasUnsavedChangesBlockIDs) {
+        hasUnsavedChangesBlockIDs[blockID] = false;
+      } else if (prevState.hasUnsavedChangesBlockIDs.hasOwnProperty(blockID)) {
+        hasUnsavedChangesBlockIDs[blockID] = prevState.hasUnsavedChangesBlockIDs[blockID];
+      } else {
+        hasUnsavedChangesBlockIDs[blockID] = false;
+      }
+      validBlockIDs[blockID] = null;
+    });
+    this.setState({ hasUnsavedChangesBlockIDs, validBlockIDs });
+  }
+
+  handleChangeHasUnsavedChanges(elementID, hasUnsavedChanges) {
+    this.setState(prevState => ({
+      hasUnsavedChangesBlockIDs: {
+        ...prevState.hasUnsavedChangesBlockIDs,
+        [elementID]: hasUnsavedChanges,
```

Comment:
Why is `[elementID]` an array? Is that even valid as a key for an object?

Reply:
Need to wrap in array so that elementID is used as a variable rather than as the actual key:

```js
const key = "3";
const prevState = { "hasUnsavedChangesBlockIDs": {"1":"a","2":"b","3":"c"} };
const someValue = {
  ...prevState.hasUnsavedChangesBlockIDs,
  [key]: "x"
};
console.log(someValue);
// => Object { 1: "a", 2: "b", 3: "x" }

const key = "3";
const prevState = { "hasUnsavedChangesBlockIDs": {"1":"a","2":"b","3":"c"} };
const someValue = {
  ...prevState.hasUnsavedChangesBlockIDs,
  key: "x"
};
console.log(someValue);
// => Object { 1: "a", 2: "b", 3: "c", key: "x" }

```

Reply:
Oh, weird, I hate it lol. Javascript is weird.